src/bsk201910/rollbacks.md

```diff
@@ -0,0 +1,407 @@
+
+When you upgrade your Pods from version 1 to version 2, the Deployment creates a new ReplicaSet and increases the count of replicas while the previous count goes to zero.
+
+**After the rolling update, the previous ReplicaSet is not deleted** — not immediately at least.
+
+Instead, it is kept around with a replicas count of 0.
+
+If you try to execute another rolling update from version 2 to version 3, you might notice that at the end of the upgrade, you have two ReplicaSets with a count of 0.
+
+_Why are the previous ReplicaSets not deleted or garbage collected?_
+
+Imagine that the current version of the container introduces a regression.
+
+You probably don't want to serve unhealthy responses to your users, so you might want to roll back to a previous version of your app.
+
+If you still have an old ReplicaSet, perhaps you could scale the current replicas to zero and increment the previous ReplicaSet count.
+
+In other words, **keeping the previous ReplicaSets around is a convenient mechanism to roll back to a previously working version of your app.**
+
+By default Kubernetes stores the last 10 ReplicaSets and lets you roll back to any of them.
+
+But you can change how many ReplicaSets should be retained by changing the `Deployment.spec.revisionHistoryLimit` in your deployment.
+
+```yaml|highlight=15|title=deployment.yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: example-deployment
+spec:
+  replicas: 3
+  selector:
+    matchLabels:
+      name: app
+  template:
+    metadata:
+      labels:
+        name: app
+    spec:
+      revisionHistoryLimit: 100
+      containers:
+      - name: app
+        image: learnk8s/hello:1.0.0
+```
+
+_What about the previous ReplicaSets?_
+
+_Could you list all the previous Replicasets that belong to a Deployment?_
+
+You can use the following command to inspect the history of your Deployment:
+
+```terminal|command=1|title=bash
+kubectl rollout history deployment/app
+```
+
+And you can rollback to a specific version with:
+
+```terminal|command=1|title=bash
+kubectl rollout undo deployment/app --to-revision=2
+```
+
+_But how does the Deployment know their ReplicaSets?_
+
+_Does it store the order in which ReplicaSets are created?_
+
+The ReplicaSets have random names with id such as `app-6ff88c4474`, so you should expect the Deployment to store a reference to them.
+
+Let's inspect the Deployment with:
+
+```terminal|command=1|title=bash
+kubectl get deployment app -o yaml
+```
+
+Nothing is looking like a list of previous 10 ReplicaSets.
+
+Deployments don't hold a reference to their ReplicaSets.
+
+At least not in the same YAML.
+
+Instead, related ReplicaSets are retrieved comparing the template section in YAML.
+
+_Remember when you learnt that Deployments are ReplicaSets with some extra features?_
+
+Kubernetes uses that information to compare Deployments and ReplicaSets and make sure that they are related.
```

Comment:
Hmm, yeah, now we are just handwaving, I would say either remove this part, or show the backpointers if you want to show how Kubernetes knows that they are "related".

Reply:
This is the part where Kubernetes finds RS based on the spec sections in the Deployment. I didn't want to go into much detail here (because it's weird and complex).

Reply:
Did you forget a link here?  Or I don't understand this reply.

I'm just against the grammar "Kubernetes uses *that* information", but then you never specify what is "that".  But I don't know how to stay concise and be less wrong, so...

Reply:
Oh, this is what you pointed out in the email. From our previous correspondence:

---

A very interesting function from kubectl:
https://github.com/kubernetes/kubernetes/blob/v1.13.7/pkg/kubectl/util/deployment/deployment.go#L156

This "findNewReplicaSet" is the function that determines which
repliacset is the "current" one for the deployment.  As you can see,
this is being done by simply comparing the pod template, so here we
already see that during normal operation it's not possible to have two
replicasets in the history that has the same pod template.  What we
also see, is that the "kubectl rollout" command family does a LOT on
the client side.  I think all this is very interesting and enlightening.